### ISSUE TITLE
`dbg_macro` suggestions are questionable

### ISSUE THREAD

**Issue by blyxyas** [rust-lang/rust-clippy#10764]
### Summary

`dbg_macro` suggests to remove the `dbg!` macros in your code. I love the idea of this lint! but seems like it just removes the `dbg!` macro. **Nothing more**.

This means that the suggestion about `dbg!(X)` will be to convert it to `X`. Does this even make sense?
imo the suggestion shouldn't be to **just** remove the `dbg!` macro leaving the contents.

### Reproducer

```rs
#[warn(clippy::dbg_macro)]
fn main() {
	dbg!("Hi");
}
```

---

This should be converted to:

```rs
#[warn(clippy::dbg_macro)]
fn main() {}
```

---

Instead, applying the suggestion does this:

```rs
#[warn(clippy::dbg_macro)]
fn main() {
	"Hi";
}
```

Which makes no sense.

### Version

```text
rustc 1.71.0-nightly (9ecda8de8 2023-04-30)
binary: rustc
commit-hash: 9ecda8de85ce893cc3fc748ab06be0b8250147a7
commit-date: 2023-04-30
host: x86_64-unknown-linux-gnu
release: 1.71.0-nightly
LLVM version: 16.0.2
```

**Comment by blyxyas** [comment]
Just checked, it doesn't error. Was checking that it warns, but with Clippy instead of running it :sweat_smile:

```
$ cargo clippy
    Checking playground v0.0.1 (/playground)
warning: statement with no effect
 --> src/main.rs:3:2
  |
3 |     "Hi";
  |     ^^^^^
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#no_effect
  = note: `#[warn(clippy::no_effect)]` on by default

warning: `playground` (bin "playground") generated 1 warning
    Finished dev [unoptimized + debuginfo] target(s) in 0.28s
```

@rustbot label: -I-suggestion-causes-error

**Comment by asquared31415** [comment]
`dbg!(some_side_effect())`, `dbg!(*thing = 0)`, and similar expressions are permitted and the contents of the macro do actually get run, always.  Additionally `dbg!` moves the value, so it's possibly meaningful for any non-copy type passed too.  I have also used it as in `if dbg!(some_cond()) {}`, where removing the whole thing would be an error, singe the result is used.

I don't think that this lint should do the work of figuring out when it should remove a whole line or just the macro call.  It's always correct to just remove the wrapping `dbg!()`, and other lints can be used afterwards that have all the same information, and can help further simplify code.

Cases where removing the inner expression is undesirable (not an exhaustive list):

```rust
fn observable() {
    println!("observable called");
}

fn main() {
    dbg!(observable());
}
```

```rust
fn main() {
    let mut data = 10_u32;
    let mut_ref = &mut data;
    dbg!(*mut_ref = 0);
    assert!(data == 0);
}
```

```rust
fn some_cond() -> bool {
    true
}

fn main() {
    let other_cond = true;
    if dbg!(some_cond()) && other_cond {}
}
```

**Comment by dswij** [comment]
> and other lints can be used afterwards that have all the same information,

As @blyxyas pointed out, it will fire `clippy::no_effect`. One strategy that might be worth considering is if we can just run the logic in `no_effect` for the current suggestion in the lint and build the suggestion from there.

If it's simple enough to refactor and reuse, it's worth the improvement in user experience for the suggestion.

**Comment by blyxyas** [comment]
Maybe just checking if there's a semicolon after the `dbg!(...)` and checking if the block inside the macro can have side effects is enough. Not getting the full `no_effect` logic.

---

### Examples:

Ignoring `"a string"`, `1` or `MyStruct::new()` if `new` doesn't have side effects.

Not ignoring `let a = 1`, `dbg!(1)` (note the lack of semicolon) or `side_effects_functions()`
